torchsde/_core/adjoint_sde.py

```diff
@@ -0,0 +1,425 @@
+                inputs=y + params,
+                grad_outputs=[-v_ * adj_y_ for v_, adj_y_ in zip(v, adj_y)],
+                allow_unused=True,
+            )
+            vjp_y, vjp_params = vjp_y_and_params[:n_tensors], vjp_y_and_params[n_tensors:]
+            vjp_params = misc.flatten(vjp_params)
+
+        return (*g_prod_eval, *vjp_y, vjp_params)
+
+    ########################################
+    #               gdg_prod               #
+    ########################################
+
+    def gdg_prod_diagonal(self, t, y_aug, v):  # For Ito/Stratonovich diagonal.
+        sde, params, n_tensors = self._base_sde, self._params, self._n_tensors
+        y, adj_y = y_aug[:n_tensors], y_aug[n_tensors:2 * n_tensors]
+
+        with torch.enable_grad():
+            y = [y_.detach().requires_grad_(True) for y_ in y]
+            adj_y = [adj_y_.detach().requires_grad_(True) for adj_y_ in adj_y]
+
+            g_eval = sde.g(-t, y)
+            gdg_times_v = misc.grad(
+                outputs=g_eval,
+                inputs=y,
+                grad_outputs=misc.seq_mul(g_eval, v),
+                allow_unused=True,
+                create_graph=True,
+            )
+            dgdy = misc.grad(
+                outputs=g_eval,
+                inputs=y,
+                grad_outputs=[torch.ones_like(y_) for y_ in y],
+                allow_unused=True,
+                create_graph=True,
+            )
+            prod_partials_adj_y_and_params = misc.grad(
+                outputs=g_eval,
+                inputs=y + params,
+                grad_outputs=misc.seq_mul(adj_y, v, dgdy),
+                allow_unused=True,
+                create_graph=True,
+            )
+            prod_partials_adj_y = prod_partials_adj_y_and_params[:n_tensors]
+            prod_partials_params = prod_partials_adj_y_and_params[n_tensors:]
+            prod_partials_params = misc.flatten(prod_partials_params)
+
+            gdg_v = misc.grad(
+                outputs=g_eval,
+                inputs=y,
+                grad_outputs=[p.detach() for p in misc.seq_mul(adj_y, v, g_eval)],
+                allow_unused=True,
+                create_graph=True
+            )
+            mixed_partials_adj_y_and_params = misc.grad(
+                outputs=gdg_v,
+                inputs=y + params,
+                grad_outputs=[torch.ones_like(p) for p in gdg_v],
+                allow_unused=True,
+            )
+            mixed_partials_adj_y = mixed_partials_adj_y_and_params[:n_tensors]
+            mixed_partials_params = mixed_partials_adj_y_and_params[n_tensors:]
+            mixed_partials_params = misc.flatten(mixed_partials_params)
+
+        return (
+            *gdg_times_v,
+            *misc.seq_sub(prod_partials_adj_y, mixed_partials_adj_y),
+            prod_partials_params - mixed_partials_params
+        )
+
+    def gdg_prod_default(self, t, y, v):  # For Ito/Stratonovich general/additive/scalar.
+        raise NotImplementedError
```

Comment:
Nitpick: can this be put before `gdg_prod_diagonal` for consistency with `g_prod`?

Reply:
Done.